kubernetes/samples/scripts/domain-lifecycle/setupPortForwarding.sh

```diff
@@ -0,0 +1,181 @@
+                                     time to wait until at least one pod is running. Default is 1m0s.
+
+    -m <kubernetes_cli>            : Kubernetes command line interface. Default is 'kubectl' if KUBERNETES_CLI env
+                                     variable is not set. Otherwise default is the value of KUBERNETES_CLI env variable.
+
+    -h                             : This help.
+   
+EOF
+exit $1
+}
+
+kubernetesCli=${KUBERNETES_CLI:-kubectl}
+domainUid=""
+domainNamespace=""
+portForwardPodName=port-forward-server
+localAddress=""
+localPort=""
+adminHost=""
+remotePort=""
+address=""
+imagePullPolicy=IfNotPresent
+imageName=alpine/socat
+waitTimeout=30
+podRunningTimeout=
+derivedAdminPort=0
+
+
+while getopts "d:l:n:m:a:p:t:w:r:h" opt; do
+  case $opt in
+    n) domainNamespace="${OPTARG}"
+    ;;
+    d) domainUid="${OPTARG}"
+    ;;
+    m) kubernetesCli="${OPTARG}"
+    ;;
+    a) localAddress="${OPTARG}";
+    ;;
+    l) localPort="${OPTARG}";
+    ;;
+    r) remotePort="${OPTARG}";
+    ;;
+    i) imageName="${OPTARG}";
+    ;;
+    p) imagePullPolicy="${OPTARG}";
+    ;;
+    s) portForwardPodName="${OPTARG}";
+    ;;
+    t) podRunningTimeout="${OPTARG}";
+    ;;
+    w) waitTimeout="${OPTARG}";
+    ;;
+    h) usage 0
+    ;;
+    *) usage 1
+    ;;
+  esac
+done
+
+#
+# Function to perform validations, read files and initialize workspace
+#
+function initialize {
+
+  # Validate the required files exist
+  validateErrors=false
+
+  validateKubernetesCliAvailable
+  validateJqAvailable
+
+  # Validate that domain name and namespace parameters are specified.
+  if [ -z "${domainUid}" ]; then
+    validationError "Please specify a domain name using '-d' e.g. '-d sample-domain1'."
+  fi
+  if [ -z "${domainNamespace}" ]; then
+    validationError "Please specify domain namespace using '-n' e.g. '-n sample-domain1-ns'."
+  fi
+
+  failIfValidationErrors
+}
+
+initialize
+
+# Get the domain in json format
+domainJson=$(${kubernetesCli} get domain ${domainUid} -n ${domainNamespace} -o json --ignore-not-found)
+if [ -z "${domainJson}" ]; then
+  printError "Unable to get domain resource for domain '${domainUid}' in namespace '${domainNamespace}'. Please make sure the 'domain_uid' and 'namespace' specified by the '-d' and '-n' arguments are correct. Exiting."
+  exit 1
+fi
+
+trap "portForwardCleanup \"${domainNamespace}\" \"$portForwardPodName\"" EXIT
+
+getAdminServerHostAdminPort "${domainUid}" "${domainNamespace}" adminHost derivedAdminPort
+if [ -z $remotePort ]; then
+remotePort=$derivedAdminPort
+fi
+
+printInfo "Creating a pod with name '$portForwardPodName' in the namespace '$domainNamespace' using image '$imageName.'"
+${kubernetesCli} run -n $domainNamespace --image-pull-policy $imagePullPolicy --restart=Never --image=$imageName $portForwardPodName -- -d -d tcp-listen:$remotePort,fork,reuseaddr tcp-connect:$adminHost:$remotePort
```

Comment:
It's possible for there to be a pod with the same name already running, perhaps because someone else is already running the script, or someone did a 'kill -9' on the script (bypassing the trap). In which case would the command fail and the script exit plus trap/delete someone else's pod?  Perhaps the script should exit early with a failure explanation if the pod is already there (without setting the trap first).